### ISSUE TITLE
tf.random.log_uniform_candidate_sampler gives undesired true class

### ISSUE THREAD

**Issue by Wang-Yu-Qing** [tensorflow/tensorflow#44758]
From tensorflow's word2vec tutorial: https://www.tensorflow.org/tutorials/text/word2vec#define_loss_function_and_compile_model

Nagative sampling is conducted using `tf.random.log_uniform_candidate_sampler`. Given the context class (true class), the goal is to sample negative classes from the whole vocabulary list. To my understanding, The negative classes must differ from the given context class. However, I found that the context class may appear in the negative classes sampled by tf.random.log_uniform_candidate_sampler. Here is the code:

```
import tensorflow as tf
SEED = 42 

# encode the words
sentence = "The wide road shimmered in the hot sun"
tokens = list(sentence.lower().split())
vocab, index = {}, 1 # start indexing from 1
vocab['<pad>'] = 0 # add a padding token 
for token in tokens:
  if token not in vocab: 
    vocab[token] = index
    index += 1
vocab_size = len(vocab)
print(vocab)
inverse_vocab = {index: token for token, index in vocab.items()}
print(inverse_vocab)


# make (hot, the) as a context pair
target_word, context_word = 6, 1
print("target: {}, context: {}".format(inverse_vocab[target_word], inverse_vocab[context_word]))


# negative sampling
# Set the number of negative samples per positive context. 
num_ns = 4

context_class = tf.reshape(tf.constant(context_word, dtype="int64"), (1, 1))
negative_sampling_candidates, _, _ = tf.random.log_uniform_candidate_sampler(
    true_classes=context_class, # class that should be sampled as 'positive'
    num_true=1, # each positive skip-gram has 1 positive context class
    num_sampled=num_ns, # number of negative context words to sample
    unique=True, # all the negative samples should be unique
    range_max=vocab_size, # pick index of the samples from [0, vocab_size]
    seed=SEED, # seed for reproducibility
    name="negative_sampling" # name of this operation
)
print("negative samples\' index: ", negative_sampling_candidates.numpy())
print("negetive samples: ", [inverse_vocab[index.numpy()] for index in negative_sampling_candidates])
# "the" will show in negative samples, if not, run it several times.
```
The word "the" is the context class of word "hot", why it could show in the sampled negative classes? Moreover, the target word hot could also be sampled as negative class. Do I misunderstand something?

**Comment by bhack** [comment]
Have you already seen https://www.tensorflow.org/extras/candidate_sampling.pdf ?

**Comment by Yu-Qing-Wang** [comment]
> Have you already seen https://www.tensorflow.org/extras/candidate_sampling.pdf ?

I've read that. If you mean 

> The random choice of S_i may or may not depend on x_i and/or T_i.

But if positive sample can show in negative samples, why the function have this `true_classes` argument? And what does the comment `# class that should be sampled as 'positive'` in the word2vec tutorial mean, what's the effect of **sampled as positive**?

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/0af7d3fc721a80b2bf6e3ae809000fa7/44758.ipynb). Thanks!

**Comment by bhack** [comment]
@Yu-Qing-Wang I am not deeply in this op but I think that `true_class` interpretation is in https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/python/ops/candidate_sampling_ops.py#L120-L128

**Comment by SarthakJShetty** [comment]
Was anyone able to fix this issue? I had the same question and would appreciate an explanation of why thr `true_classes` is included in the distribution returned.

Thanks!

**Comment by bhack** [comment]
@MarkDaoust Do you think that we could improve the documentation here?

**Comment by eshu99** [comment]
Facing the same issue. Does using tf.nn.fixed_unigram_candidate_sampler instead of tf.random.log_uniform_candidate_sampler for negative sampling not cause this issue of generating undesired true class also?

**Comment by SarthakJShetty** [comment]
> Facing the same issue. Does using tf.nn.fixed_unigram_candidate_sampler instead of tf.random.log_uniform_candidate_sampler for negative sampling not cause this issue of generating undesired true class also?

From what I understand there's a probability attached to each label while sampling. So there is a chance that the `true_label` gets returned as a false label while generating the context classes here. I might be wrong here, but the PDFs and the code snippet referenced above point in that direction.

**Comment by leondgarse** [comment]
- I think this is designed to be, but the line `pass the context word as true class to exclude it from being sampled` in the tutorial is really misguiding. I'm just learning this from the tutorial, so there maybe mistakes in my understanding.
- In my test, `true_classes` has no effect on the sampled results, seems it's only used to calculate the second returned value `true_expected_count`:
  ```py
  # Do sampling 1000 times using true_classes [0, 8]
  sample_func = lambda ii: tf.random.log_uniform_candidate_sampler(true_classes=[[ii]], num_true=1, num_sampled=4, unique=True, range_max=8, seed=42)
  dd = {ii : np.stack([sample_func(ii)[0].numpy() for jj in range(1000)]) for ii in range(8)}
  # Calculate the distribution in each true_class
  for ii in dd:
      print("true_class:", ii, ", negative value_counts:", pd.value_counts(dd[ii].flatten()).to_dict())
  # true_class: 0 , negative value_counts: {0: 871, 1: 722, 2: 584, 3: 466, 4: 402, 5: 329, 7: 319, 6: 307}
  # true_class: 1 , negative value_counts: {0: 867, 1: 695, 2: 571, 3: 485, 4: 411, 5: 380, 6: 316, 7: 275}
  # true_class: 2 , negative value_counts: {0: 869, 1: 716, 2: 541, 3: 488, 4: 389, 5: 357, 6: 321, 7: 319}
  # true_class: 3 , negative value_counts: {0: 877, 1: 715, 2: 582, 3: 482, 4: 394, 5: 355, 6: 318, 7: 277}
  # true_class: 4 , negative value_counts: {0: 883, 1: 716, 2: 566, 3: 489, 4: 394, 5: 367, 6: 316, 7: 269}
  # true_class: 5 , negative value_counts: {0: 862, 1: 717, 2: 583, 3: 496, 4: 376, 5: 357, 6: 315, 7: 294}
  # true_class: 6 , negative value_counts: {0: 859, 1: 725, 2: 575, 3: 482, 4: 413, 5: 356, 6: 302, 7: 288}
  # true_class: 7 , negative value_counts: {0: 880, 1: 724, 2: 555, 3: 488, 4: 425, 5: 324, 7: 302, 6: 302}
  
  # Result of `true_expected_count`
  print({ii : np.mean([sample_func(ii)[1].numpy() for jj in range(1000)]) for ii in range(8)})
  # {0: 0.99967235, 1: 0.7245632, 2: 0.5737029, 3: 0.47004792, 4: 0.3987442, 5: 0.34728608, 6: 0.3084587, 7: 0.27554017}
  ```
- Simmilar result with `tf.nn.fixed_unigram_candidate_sampler` or `tf.keras.preprocessing.sequence.skipgrams` with `negative_samples=4`:
  ```py
  unigrams = [0.99967235, 0.7245632, 0.5737029, 0.47004792, 0.3987442, 0.34728608, 0.3084587, 0.27554017]
  sample_func = lambda ii: tf.nn.fixed_unigram_candidate_sampler(true_classes=[[ii]], num_true=1, num_sampled=4, unique=True, range_max=8, unigrams=unigrams)
  dd = {ii : np.stack([sample_func(ii)[0].numpy() for jj in range(1000)]) for ii in range(8)}
  print("true_class:", 0, "value_counts:", pd.value_counts(dd[0].flatten()).to_dict())
  print("true_class:", 7, ", negative value_counts:", pd.value_counts(dd[7].flatten()).to_dict())
  # true_class: 0 , negative value_counts: {0: 793, 1: 666, 2: 546, 3: 475, 4: 426, 5: 382, 7: 356, 6: 356}
  # true_class: 7 , negative value_counts: {0: 759, 1: 674, 2: 579, 3: 526, 4: 418, 5: 386, 6: 356, 7: 302}
  ```
- That means, **if `0` representing the word `the` and it's a `context_word`, which would be a high probability, it also will be selected as a `negative sample` with high probability no matter what the `context_word` is**.
- But in the original [word2vec TrainModelThread](https://github.com/svn2github/word2vec/blob/99e546e27cae10aa20209dae1ed98716ac9022e9/word2vec.c#L516) I think the positive word is actually excluded in negatives. In my understanding, I think current behavior of `log_uniform_candidate_sampler` is better.
- Actually in a real practice like the `shakespeare` dataset, the ratio selecting a positive as negative in the same sample is low, so this behavior may doesn't matter:
  ```py
  # After generated `targets, contexts, labels` following the tutorial
  rr_contexts = np.array(contexts)[:, :, 0]
  rr = [rr_contexts[ii][0] in rr_contexts[ii][1:] for ii in range(rr_contexts.shape[0])]
  print("Total negatives containing positive:", np.sum(rr), ", ratio:", np.sum(rr) / rr_contexts.shape[0])
  # Total negatives containing positive: 2226 , ratio: 0.03430843685459758
  print("Sample:", rr_contexts[np.array(rr)][:5].tolist())
  # Sample: [[1, 3, 0, 73, 1], [1, 1, 2, 47, 625], [4, 9, 717, 11, 4], [8, 15, 37, 26, 8], [1, 97, 1, 4, 120]]
  ```
- Moreover, it's a high ratio that a positive pair `(target_word, context_word)` would also be a negative sample pair:
  ```py
  targets, contexts, labels = np.array(targets), np.array(contexts), np.array(labels)
  dd = pd.DataFrame({"targets": targets.tolist(), "pos": contexts[:, 0, 0].tolist(), "neg": contexts[:, 1:, 0].tolist()})
  cc = dd.groupby('targets').apply(lambda ii: np.sum([jj in ii['pos'].values for jj in np.concatenate(ii['neg'].values)]))
  print("Total negative pairs containing positive pair:", cc.sum(), "ratio:", cc.sum() / dd.shape[0])
  # Total negative pairs containing positive pair: 38660 ratio: 0.5953095887035925
  ```
- So IMO, we need an update on the tutorial description, really misguiding.

**Comment by wangpengmit** [comment]
According to the big table in https://www.tensorflow.org/extras/candidate_sampling.pdf , the phrase "negative samples" may mean different things in different algorithms. For example, in NCE "negative samples" means S_i (which is just the sampled classes) which may overlap with true classes, while in Sampled Logistic "negative samples" means S_i - T_i which excludes the true classes.

For `tf.random.log_uniform_candidate_sampler`, the return value `sampled_candidates` corresponds to S_i, not any specific definition of "negative samples" in any specific algorithm. It's the caller's responsibility to pick an algorithm and calculate the "negative samples" defined by that algorithm (e.g. S_i - T_i).

The `true_classes` argument is for calculating the `true_expected_count` output (as a by-product of the op's main calculation), which may be needed by some algorithms (according to https://www.tensorflow.org/extras/candidate_sampling.pdf). It's not for excluding true classes in the return value `sampled_candidates` (again that step is algorithm-specific and should be carried out by the caller).

**Comment by Saduf2019** [comment]
@Wang-Yu-Qing 
Is this still an issue

**Comment by gqqnbig** [comment]
Even if Wang-Yu-Qing is fine, I think we need to update documentations to make clear the meaning of "negative samples".

**Comment by wangpengmit** [comment]
Agreed. We'll track the documentation update in #49490.

**Comment by gqqnbig** [comment]
In that case, I will close this issue and look at  #49490.

**Comment by Saduf2019** [comment]
In that case can this be move to closed status.

**Comment by wangpengmit** [comment]
Closing this issue and tracking the documentation update in #49490. Feel free to reopen.

**Comment by viralbthakar** [comment]
> According to the big table in https://www.tensorflow.org/extras/candidate_sampling.pdf , the phrase "negative samples" may mean different things in different algorithms. For example, in NCE, "negative samples" means S_i (which is just the sampled classes) which may overlap with true classes, while in Sampled Logistic "negative samples" means S_i - T_i which excludes the true classes.
> 
> For `tf.random.log_uniform_candidate_sampler`, the return value `sampled_candidates` corresponds to S_i, not any specific definition of "negative samples" in any specific algorithm. It's the caller's responsibility to pick an algorithm and calculate the "negative samples" defined by that algorithm (e.g. S_i - T_i).
> 
> The `true_classes` argument is for calculating the `true_expected_count` output (as a by-product of the op's main calculation), which may be needed by some algorithms (according to https://www.tensorflow.org/extras/candidate_sampling.pdf). It's not for excluding true classes in the return value `sampled_candidates` (again that step is algorithm-specific and should be carried out by the caller).

@wangpengmit 

In [this](https://www.tensorflow.org/tutorials/text/word2vec#skip-gram_and_negative_sampling) section the idea of negative sample is defined different from what you are highlighting in the comment. 

> A negative sample is defined as a (target_word, context_word) pair such that the context_word does not appear in the window_size neighborhood of the target_word. For the example sentence, these are a few potential negative samples (when window_size is 2).

Based on this do you think we should explicitly exclude the context class from negative sampling?
